### PR TITLE
Menu and ContextMenu: add ScrollView

### PR DESCRIPTION
**Pull request by iBat**

### CODE REVIEW

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 1]
```diff
@@ -603,6 +605,24 @@ class ContextMenu extends MenuBase {
         }
     }
 
+    _initScrollView(container) {
+        const maxHeight = this._getMaxHeight();
+        const containerHeight = getOuterHeight(container);
+
+        container.css('position', 'fixed');
+        container.css('height', Math.min(containerHeight, maxHeight));
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [code style]
Please don't pass calculated expressions as parameters. It's better to move it to a variable for easier debugging.
Please, update it in all such places

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 2]
```diff
@@ -860,7 +882,13 @@ class ContextMenu extends MenuBase {
 
             this._setOptionWithoutOptionChange('visible', true);
             this._overlay.option('position', position);
-            promise = this._overlay.show();
+            promise = this._overlay.show()
+                .then(() => {
+                    const $subMenu = $(this._overlay.content()).children('.dx-submenu');
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [code style]
```suggestion
                    const $subMenu = $(this._overlay.content()).children(`.${DX_SUBMENU_CLASS}`);
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js [hunk 3]
```diff
@@ -211,7 +213,8 @@ QUnit.module('Rendering', moduleConfig, () => {
 
         instance.option('items', [{ text: 3 }, { text: 4 }]);
         instance.show();
-        assert.equal($('.dx-overlay').length, 1, 'only one overlay should exists');
+        this.clock.tick(0);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js:218]: [unnecessary changes]
I skipped the change and test is green.

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js [hunk 4]
```diff
@@ -267,6 +270,29 @@ QUnit.module('Rendering', moduleConfig, () => {
...
+    QUnit.test('height of the context menu should be limited', function(assert) {
+        new ContextMenu(this.$element, {
+            items: (new Array(99)).fill(null).map((_, idx) => ({ text: idx })),
+            visible: true,
+        });
+        this.clock.tick(0);
+
+        const $submenu = $(`.${DX_SUBMENU_CLASS}`);
+
+        assert.ok($submenu.find(`.${DX_SCROLLVIEW_CONTENT_CLASS}`).height() > $(window).height(), 'total height of submenu is exceeds the window');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js]:
```suggestion
        assert.ok($submenu.find(`.${DX_SCROLLVIEW_CONTENT_CLASS}`).height() > $(window).height(), 'total height of submenu exceeds the window');
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js [hunk 5]
```diff
@@ -267,6 +270,29 @@ QUnit.module('Rendering', moduleConfig, () => {
...
+        assert.ok($submenu.height() < $(window).height(), 'height of submenu wrapper is not exceeds the window');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js]:
```suggestion
        assert.ok($submenu.height() < $(window).height(), 'height of submenu wrapper does not exceed the window height');
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js [hunk 6]
```diff
@@ -287,7 +313,8 @@ QUnit.module('Showing and hiding context menu', moduleConfig, () => {
         const instance = new ContextMenu(this.$element, { items: [{ text: 1 }], visible: true });
 
         instance.option('items', [{ text: 1 }]);
-        assert.equal($('.dx-overlay').length, 1, 'overlays cleaned correctly');
+        this.clock.tick(0);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js:416]: [unnecessary change]
same here

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 7]
```diff
@@ -380,6 +384,47 @@ QUnit.module('Menu rendering', {
...
+        });
+
+        const $item1 = $(menu.element).find(`.${DX_MENU_ITEM_CLASS}`).eq(0);
+
+        $($item1).trigger('dxclick');
+        this.clock.tick(0);
+
+        const $submenu = $(`.${DX_SUBMENU_CLASS}`);
+
+        assert.ok($submenu.find(`.${DX_SCROLLVIEW_CONTENT_CLASS}`).height() > $(window).height(), 'total height of submenu is exceeds the window');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]:
```suggestion
        assert.ok($submenu.find(`.${DX_SCROLLVIEW_CONTENT_CLASS}`).height() > $(window).height(), 'total height of submenu exceeds the window height');
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 8]
```diff
@@ -380,6 +384,47 @@ QUnit.module('Menu rendering', {
...
+        assert.ok($submenu.height() < $(window).height(), 'height of submenu wrapper is not exceeds the window');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]:
```suggestion
        assert.ok($submenu.height() < $(window).height(), 'height of submenu wrapper does not exceed the window height');
```

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 9]
```diff
@@ -603,6 +605,24 @@ class ContextMenu extends MenuBase {
...
+        const maxHeight = this._getMaxHeight();
+        const containerHeight = getOuterHeight(container);
+
+        container.css('position', 'fixed');
+        container.css('height', Math.min(containerHeight, maxHeight));
+        this._createComponent(container, ScrollView, {});
+    }
+
+    _getMaxHeight() {
+        const $element = this.$element();
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [not working code]
As far as I see this code is totally not working.
1. `$element` here is a root element of contextMenu => `getOuterHeight($element)` is always `0` 
2. and `$element.offset().top` shows not a contextMenu's left top corner but just a location of element on which we created an instance.

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 10]
```diff
@@ -603,6 +605,24 @@ class ContextMenu extends MenuBase {
...
+        const offsetTop = $element.offset().top;
+        const windowHeight = getOuterHeight(window);
+        const maxHeight = Math.max(offsetTop, windowHeight - offsetTop - getOuterHeight($element));
+
+        return Math.min(windowHeight, maxHeight);
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [useless code]
We calculate `maxHeight` as `windowHeight - offsetTop - getOuterHeight($element)` => it is never greater than `windowHeight`

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 11]
```diff
@@ -622,6 +642,8 @@ class ContextMenu extends MenuBase {
         if(!this._isSubmenuVisible($submenu)) {
             this._drawSubmenu($item);
         }
+
+        this._initScrollView(isSubmenuRendered ? $submenu : $item.children(`.${DX_SUBMENU_CLASS}`));
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [question, possibly useless code]
How submenu can be not rendered here if we called `_renderSubmenuItems` and `_drawSubmenu` above?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js [hunk 12]
```diff
@@ -267,6 +270,29 @@ QUnit.module('Rendering', moduleConfig, () => {
...
+            items: (new Array(99)).fill(null).map((_, idx) => ({ text: idx })),
+            visible: true,
+        });
+        this.clock.tick(0);
+
+        const $submenu = $(`.${DX_SUBMENU_CLASS}`);
+
+        assert.ok($submenu.find(`.${DX_SCROLLVIEW_CONTENT_CLASS}`).height() > $(window).height(), 'total height of submenu is exceeds the window');
+        assert.ok($submenu.height() < $(window).height(), 'height of submenu wrapper is not exceeds the window');
+    });
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js:395]: [code coverage]
These tests are not enough. In case our calculations (for example, as in `getMaxHeight`) are incorrect we don't have red tests at the moment. **Each meaningful action in code should be tested.**

Also I don't see  checks for submenus (remove changes in `_showSubmenu` and everything is green).

File: packages/devextreme/js/ui/context_menu/ui.menu_base.js [hunk 13]
```diff
@@ -400,36 +400,37 @@ class MenuBase extends HierarchicalCollectionWidget {
     }
 
     _renderItems(nodes, submenuContainer) {
-        if(nodes.length) {
-            this.hasIcons = false;
+        if(!nodes.length) {
```
Comment (iBat) [inline packages/devextreme/js/ui/context_menu/ui.menu_base.js:403]: Unnecessary in this PR but code of this method will be changed in next PRs related to this feature

File: apps/demos/testing/widgets/contextmenu/ContextMenuShare.test.js [hunk 14]
```diff
@@ -6,7 +6,7 @@ import { testScreenshot } from '../../../utils/visual-tests/helpers/theme-utils'
 const IMAGE_ID = 'image';
 const MENU_ITEM_TEXT_CLASS = 'dx-menu-item-text';
 
-fixture('ContextMenu.Templates')
+fixture.skip('ContextMenu.Templates')
```
Comment (iBat) [inline apps/demos/testing/widgets/contextmenu/ContextMenuShare.test.js:9]: All skipped tests will be fixed in next PRs for this feature

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 15]
```diff
@@ -860,7 +882,13 @@ class ContextMenu extends MenuBase {
 
             this._setOptionWithoutOptionChange('visible', true);
             this._overlay.option('position', position);
-            promise = this._overlay.show();
+            promise = this._overlay.show()
+                .then(() => {
+                    const $subMenu = $(this._overlay.content()).children('.dx-submenu');
+                    if($subMenu.length) {
+                        this._initScrollView($subMenu);
```
Comment (iBat) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: for 1st level of submenus in Menu and 1st level of ContextMenu.

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js (hunk 11 above)
Comment (iBat) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: For 2nd and following levels of Menu and ContextMenu

File: packages/devextreme/js/ui/context_menu/ui.menu_base.js (hunk 13 above)
Comment (ksercs) [reply packages/devextreme/js/ui/context_menu/ui.menu_base.js:403]: Anyway it looks much better :) New option is just easier to read

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js (hunk 3 above)
Comment (iBat) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js:218]: It will be red in jQuery mode.

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js (hunk 10 above)
Comment (iBat) [reply packages/devextreme/js/ui/context_menu/ui.context_menu.js]: offset can be negative

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js (hunk 6 above)
Comment (iBat) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js:416]: It is necessary only in jQuery mode

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 16]
```diff
@@ -603,25 +605,46 @@ class ContextMenu extends MenuBase {
...
         if(!isSubmenuRendered) {
             this._renderSubmenuItems(node, $item);
+            $submenu = $item.children(`.${DX_SUBMENU_CLASS}`);
         }
 
         if(!this._isSubmenuVisible($submenu)) {
             this._drawSubmenu($item);
         }
+
+        this._initScrollView($submenu);
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: why do u pass submenu here? The function's parameter is called "container". Submenu is not a container, or it's about exactly container for scrollview? Then I think it's better to select another name.

Your solution still does not work. Just click to open contextMenu and see that submenu takes window's height because offset if 0, click one more time - height is reduced

**Review by ksercs — CHANGES_REQUESTED** [review]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js [hunk 17]
```diff
@@ -267,6 +270,53 @@ QUnit.module('Rendering', moduleConfig, () => {
...
+            visible: true,
+        });
+        this.clock.tick(0);
+
+        const $submenu = $(`.${DX_SUBMENU_CLASS}`);
+
+        assert.ok($submenu.find(`.${DX_SCROLLVIEW_CONTENT_CLASS}`).height() > $(window).height(), 'total height of submenu exceeds the window');
+        assert.strictEqual($submenu.outerHeight(), $(window).height(), 'menu uses the full height of the window');
+        assert.strictEqual($submenu.offset().top, 0, 'menu does not cross the window border');
+    });
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js:395]: I see this is the test for root contextMenu opening.
What about submenu? Should it also take the whole screen, or it should take only space above/below parent item?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 18]
```diff
@@ -380,6 +384,82 @@ QUnit.module('Menu rendering', {
...
+
+        const $item1 = $(menu.element).find(`.${DX_MENU_ITEM_CLASS}`).eq(0);
+
+        $($item1).trigger('dxclick');
+        this.clock.tick(0);
+
+        const $submenu = $(`.${DX_SUBMENU_CLASS}`);
+
+        assert.ok($submenu.find(`.${DX_SCROLLVIEW_CONTENT_CLASS}`).height() > $(window).height(), 'total height of submenu exceeds the window height');
+        assert.strictEqual($submenu.outerHeight(), $(window).height(), 'menu uses the full height of the window');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]: hm, why should it take the whole height of the window?
I believe it should only take the place below the clicked item.

Also i think scenario for 2nd level submenu should be tested as well.

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 19]
```diff
@@ -380,6 +385,118 @@ QUnit.module('Menu rendering', {
...
+        const $item1 = $(menu.element).find(`.${DX_MENU_ITEM_CLASS}`).eq(0);
+
+        $($item1).trigger('dxclick');
+        this.clock.tick(0);
+
+        const $submenu = $(`.${DX_SUBMENU_CLASS}`);
+
+        assert.ok($submenu.find(`.${DX_SCROLLVIEW_CONTENT_CLASS}`).height() > $(window).height(), 'total height of submenu exceeds the window height');
+        assert.strictEqual($submenu.offset().top, $item1.offset().top + $item1.outerHeight(), 'submenu aligned to a clicked item');
+        assert.strictEqual($submenu.outerHeight(), $(window).height() - $item1.offset().top - $item1.outerHeight(), 'menu uses all available space');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]: [unstable test]
such checks should use `roughEqual`. 
For example, this test is red in my machine

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 20]
```diff
@@ -380,6 +385,118 @@ QUnit.module('Menu rendering', {
...
+
+        const $menuItem = $($(submenu._overlay.content()).find(`.${DX_MENU_ITEM_CLASS}`).first());
+        $(submenu.itemsContainer()).trigger($.Event('dxhoverstart', { target: $menuItem.get(0) }));
+        $($menuItem).trigger('dxpointermove');
+        this.clock.tick(0);
+        const $submenu = $($(submenu._overlay.content()).find(`.${DX_SUBMENU_CLASS}`).eq(1));
+        assert.ok($submenu.hasClass(DX_SCROLLVIEW_CLASS), 'ScrollView initialized on nested submenu');
+    });
+
+    QUnit.test('height of the submenu should be correctly positioned', function(assert) {
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]: [test naming]
smth wrong with this test name. "height should be positioned"?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 21]
```diff
@@ -380,6 +385,118 @@ QUnit.module('Menu rendering', {
...
+
+        $(submenu.itemsContainer()).trigger($.Event('dxhoverstart', { target: $menuItem.get(0) }));
+        $($menuItem).trigger('dxpointermove');
+        this.clock.tick(0);
+
+        const $nestedSubmenu = $($(submenu._overlay.content()).find(`.${DX_SUBMENU_CLASS}`).eq(1));
+        const $nestedItemsContainer = $nestedSubmenu.find(`.${DX_CONTEXT_MENU_ITEMS_CONTAINER_CLASS}`).eq(0);
+
+        assert.strictEqual($nestedItemsContainer.offset().top, $menuItem.offset().top, 'Nested submenu positioned to a clicked item');
+        assert.strictEqual($nestedSubmenu.outerHeight(), $(window).height() - $nestedSubmenu.offset().top - 1, 'Nested submenu uses all available space');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]: [refactoring]
I suggest u moving `1` to a constant with meaningful name to show it's not a magic number

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 22]
```diff
@@ -603,25 +605,67 @@ class ContextMenu extends MenuBase {
         }
     }
 
+    _initScrollView($container, anchor, considerAnchorHeight = true) {
+        const containerHeight = getOuterHeight($container);
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [broken behavior]
why do you use container height here, not content height?
when u open submenu second time container height is already restricted => new calculations can be incorrect

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 23]
```diff
@@ -603,25 +605,67 @@ class ContextMenu extends MenuBase {
...
         if(!isSubmenuRendered) {
             this._renderSubmenuItems(node, $item);
+            $submenu = $item.children(`.${DX_SUBMENU_CLASS}`);
         }
 
         if(!this._isSubmenuVisible($submenu)) {
             this._drawSubmenu($item);
         }
+
+        this._initScrollView($submenu, $item, false);
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [clear logic, refactoring]
Do we really need to init a scrollView on each opening? Seems like it can be rendered once on first rendering and then we should only update height, no?

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 24]
```diff
@@ -603,25 +605,67 @@ class ContextMenu extends MenuBase {
...
+        const maxHeight = this._getMaxHeight(containerHeight, anchor, considerAnchorHeight);
+        const menuHeight = Math.min(containerHeight, maxHeight);
+        const cssProps = {
+            height: menuHeight,
+            overflow: 'visible',
+        };
+        const containerOffset = $container.offset().top;
+
+        if(containerOffset < 0) {
+            cssProps.top = -containerOffset;
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [strange code]
I understand what is this for. But:
- I commented these rows and tests are green;
- it's not ok what we affect positioning in `initScrollView` function;
- it's not very good that we have 2 contradicting css props `transform` and `top` on the element;

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 25]
```diff
@@ -603,25 +605,67 @@ class ContextMenu extends MenuBase {
...
+        }
+        $container.css(cssProps);
+        this._createComponent($container, ScrollView, {});
+    }
+
+    _getMaxHeight(containerHeight, anchor, considerAnchorHeight) {
+        const windowHeight = getOuterHeight(window);
+        const isContextMenuRootEvent = anchor.type === DEFAULT_SHOW_EVENT || !anchor.offset;
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [refactoring, strange code]
If here u're trying to check that anchor is event let's do it directly
```javascript
const isEventAnchor = isEvent(anchor);
```

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 26]
```diff
@@ -603,25 +605,67 @@ class ContextMenu extends MenuBase {
...
+        const document = domAdapter.getDocument();
+        const isAnchorDocument = anchor.length && anchor[0] === document;
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js:625]: [refactoring]
We can use an existing util as well
```suggestion
        const isAnchorDocument = isDocument(anchor);
```

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 27]
```diff
@@ -603,25 +605,67 @@ class ContextMenu extends MenuBase {
...
+
+        if(isContextMenuRootEvent || isAnchorDocument) {
+            return windowHeight;
+        }
+
+        const offsetTop = anchor.offset().top;
+        let maxHeight = Math.max(offsetTop, windowHeight - offsetTop);
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [mistake]
it's not correct to make `max` here and plus/minus anchor height later because this `+/-` can affect the result of this `max`. 
For example, now Menu's flipped 1st-level submenu height is incorrect

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 28]
```diff
@@ -380,6 +385,118 @@ QUnit.module('Menu rendering', {
...
+    });
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js:590]: [test coverage]
Scenarios for flipping are missed

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 29]
```diff
@@ -603,22 +606,58 @@ class ContextMenu extends MenuBase {
...
+        const $itemsContainer = $container.find(`.${DX_MENU_ITEMS_CONTAINER_CLASS}`);
+        const contentHeight = getOuterHeight($itemsContainer);
+        const maxHeight = this._getMaxHeight(anchor, !isNestedSubmenu);
+        const menuHeight = Math.min(contentHeight, maxHeight);
+
+        $container.css({
+            height: isNestedSubmenu ? menuHeight : '100%',
+            overflow: 'visible',
+        });
+        if(!$container.hasClass(SCOLLVIEW_CLASS)) {
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [possibly useless code]
Is it necessary? Seems like if we try to recreate ScrollMenu on element where it's already created, nothing is happened

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 30]
```diff
@@ -603,22 +606,58 @@ class ContextMenu extends MenuBase {
...
+            return windowHeight;
+        }
+
+        const offsetTop = getOffsetTop(anchor[0]);
+        const anchorHeight = getOuterHeight(anchor);
+        const maxHeight = considerAnchorHeight
+            ? Math.max(offsetTop, windowHeight - offsetTop - anchorHeight)
+            : Math.max(offsetTop + anchorHeight, windowHeight - offsetTop);
+
+        return Math.min(windowHeight, maxHeight);
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [possibly useless code]
Is it necessary now?

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 31]
```diff
@@ -603,22 +606,58 @@ class ContextMenu extends MenuBase {
         }
     }
 
+    _initScrollView($container, anchor, isNestedSubmenu) {
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [code splitting]
I suggest u splitting this function to several with minimized responsibility.
`_initSrollView` should be called only of overlay/submenu first rendering and only init scrollview.
Function with name like `_setSubMenuHeight` can update height and should be called every time it's necessary.
` overflow: 'visible'` can be set in styles I believe

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 32]
```diff
@@ -603,22 +606,58 @@ class ContextMenu extends MenuBase {
...
         const isSubmenuRendered = $submenu.length;
 
         super._showSubmenu($item);
 
         if(!isSubmenuRendered) {
             this._renderSubmenuItems(node, $item);
+            $submenu = $item.children(`.${DX_SUBMENU_CLASS}`);
         }
 
+        this._initScrollView($submenu, $item, true);
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: I guess we should init scrollview on submenu creation (where we create an appropriate element).
Here we should only update height (or somewhere near, check it plz)

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 33]
```diff
@@ -859,7 +898,21 @@ class ContextMenu extends MenuBase {
...
+                return this._getMaxHeight(position.of);
+            });
+            this._overlay.option('maxHeight', () => {
+                const $subMenu = $(this._overlay.content()).children(`.${DX_SUBMENU_CLASS}`);
+                const $content = $subMenu.find(`.${DX_MENU_ITEMS_CONTAINER_CLASS}`);
+                return getOuterHeight($content);
+            });
             this._overlay.option('position', position);
+
+            const $subMenu = $(this._overlay.content()).children(`.${DX_SUBMENU_CLASS}`);
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [copypaste]

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 34]
```diff
@@ -859,7 +898,21 @@ class ContextMenu extends MenuBase {
...
             this._setOptionWithoutOptionChange('visible', true);
+            this._overlay.option('height', () => {
+                return this._getMaxHeight(position.of);
+            });
+            this._overlay.option('maxHeight', () => {
+                const $subMenu = $(this._overlay.content()).children(`.${DX_SUBMENU_CLASS}`);
+                const $content = $subMenu.find(`.${DX_MENU_ITEMS_CONTAINER_CLASS}`);
+                return getOuterHeight($content);
+            });
             this._overlay.option('position', position);
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [code style]
all 3 options can be updated in patch

File: packages/devextreme/js/ui/context_menu/ui.context_menu.js [hunk 35]
```diff
@@ -967,6 +1020,15 @@ class ContextMenu extends MenuBase {
     }
 }
 
+function getOffsetTop(el) {
```
Comment (ksercs) [inline packages/devextreme/js/ui/context_menu/ui.context_menu.js]: [possibly excess code]
I don't see this util is useful.
Cannot we just use `getBoundingClientRect`?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js [hunk 36]
```diff
@@ -211,7 +215,8 @@ QUnit.module('Rendering', moduleConfig, () => {
 
         instance.option('items', [{ text: 3 }, { text: 4 }]);
         instance.show();
-        assert.equal($('.dx-overlay').length, 1, 'only one overlay should exists');
+        this.clock.tick(0);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/contextMenu.tests.js:218]: [possibly useless code]
we got rid of `promise.then` => maybe these ticks are already not necessary?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 37]
```diff
@@ -380,6 +386,199 @@ QUnit.module('Menu rendering', {
...
+        $($menuItem).trigger('dxpointermove');
+        this.clock.tick(0);
+
+        const $nestedSubmenu = $($(submenu._overlay.content()).find(`.${DX_SUBMENU_CLASS}`).eq(1));
+        const $nestedItemsContainer = $nestedSubmenu.find(`.${DX_CONTEXT_MENU_ITEMS_CONTAINER_CLASS}`).eq(0);
+
+        assert.roughEqual($nestedItemsContainer.offset().top, $menuItem.offset().top, .1, 'Nested submenu positioned to a clicked item');
+        assert.roughEqual(
+            $nestedSubmenu.outerHeight(), $(window).height() - $nestedSubmenu.offset().top - BORDER_WIDTH,
+            .1, 'Nested submenu uses all available space');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]: [code style]
fix indents pls
```javascript
assert.roughEqual(
   first_param,
   second_param,
   third_param
);
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 38]
```diff
@@ -380,6 +386,199 @@ QUnit.module('Menu rendering', {
...
+            showFirstSubmenuMode: 'onClick',
+            showSubmenuMode: { name: 'onHover', delay: 0 },
+        });
+        const $item1 = $(menu.element).find(`.${DX_MENU_ITEM_CLASS}`).eq(0);
+
+        $($item1).trigger('dxclick');
+        this.clock.tick(0);
+
+        const submenu = getSubMenuInstance($item1);
+        this.clock.tick(MOUSETIMEOUT);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]: [possibly useless code]
why is it necessary?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 39]
```diff
@@ -380,6 +386,199 @@ QUnit.module('Menu rendering', {
...
+                        text: 'item 111',
+                    }],
+                }],
+            }],
+            showFirstSubmenuMode: 'onClick',
+            showSubmenuMode: { name: 'onHover', delay: 0 },
+        });
+        const $item1 = $(menu.element).find(`.${DX_MENU_ITEM_CLASS}`).eq(0);
+
+        $($item1).trigger('dxclick');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]: [possibly useless]
do we really need additional `$` wrap?
I see it in many places

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 40]
```diff
@@ -380,6 +386,199 @@ QUnit.module('Menu rendering', {
...
+        const menu = createMenuInWindow({
+            items: [{
+                text: 'item 1',
+                items: (new Array(99)).fill(null).map((_, idx) => ({ text: idx })),
+            }],
+            showFirstSubmenuMode: 'onClick',
+        });
+        const $item1 = $(menu.element).find(`.${DX_MENU_ITEM_CLASS}`).eq(0);
+
+        menu.element.css('top', 10000 + $(window).height() - 50);
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js]: [refactoring]
Let's move it to variables to make the code clearer:
```javascript
const FIXTURE_OFFSET = 10000;
const menuRootOffset =  $(window).height() - 50;
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 41]
```diff
@@ -380,6 +386,199 @@ QUnit.module('Menu rendering', {
...
+        if(!isDeviceDesktop(assert)) {
+            return;
+        }
+
+        const menu = createMenuInWindow({
+            items: [{
+                text: 'item 1',
+                items: [{
+                    text: 'item 11',
+                    items: (new Array(99)).fill(null).map((_, idx) => ({ text: idx })),
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js:567]:
```suggestion
                    items: new Array(99)).fill(null).map((_, idx) => ({ text: idx }),
```

Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js:567]: [tests splittin idea]
Don't u wanna to union all these tests into single module to unify some things like items creation, submenu getting function, etc?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js [hunk 42]
```diff
@@ -60,6 +63,9 @@ const CLICKTIMEOUT = 51;
 const ANIMATION_TIMEOUT = 100;
 const MENU_ITEM_WIDTH = 100;
 const MOUSETIMEOUT = 50;
+const BORDER_WIDTH = 1;
+const FIXTURE_OFFSET = 10000;
+const menuRootOffset = $(window).height() - 50;
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js:68]: I mean u use this constant in the module, not declare it for the whole file :)

**Review by ksercs — DISMISSED** [review]

**Review by pomahtri — DISMISSED** [review]

**Review by pomahtri — APPROVED** [review]

**Review by ksercs — APPROVED** [review]
